File: packages/nxuskit-py/src/nxuskit/license.py
"""License management for the nxusKit SDK.

Safe Python wrappers over the C ABI license functions: token resolution,
validation, and machine fingerprinting.
"""

from __future__ import annotations

import json
from dataclasses import dataclass, field
from typing import Dict, List, Optional

from nxuskit._ffi import ffi, last_error, lib


@dataclass
class LicenseResolution:
    """Token resolution result from the license precedence chain."""

    source: str
    """Where the token was found: 'env_var', 'file', 'api_param', or 'none'"""
    token_type: str
    """Token type: 'trial', 'developer', 'deployment', or 'none'"""
    valid: bool
    """Whether the token passed validation"""
    error: Optional[str]
    """Error message if validation failed"""
    product_id: Optional[str] = field(default=None)
    """Product identifier (e.g., 'nxuskit')"""
    effective_limits: Optional[Dict] = field(default=None)
    """Resolved numerical limits (catalog defaults + token overrides)"""
    features: Optional[List[str]] = field(default=None)
    """Effective feature list for the resolved edition"""


@dataclass
class TokenInfo:
    """Token validation result."""

    valid: bool
    """Whether the token is valid"""
    token_type: str
    """Token type: 'trial', 'developer', 'deployment'"""
    edition: Optional[str]
    """Edition granted by the token"""
    days_remaining: Optional[int]
    """Days until token expiry (None for deployment tokens)"""
    error: Optional[str]
    """Error message if validation failed"""
    result: str
    """Entitlement result code"""


def license_resolve(explicit_key: Optional[str] = None) -> LicenseResolution:
    """Resolve the active license token from all available sources.

    Resolution order:
        1. ``NXUSKIT_LICENSE_TOKEN`` environment variable
        2. ``~/.nxuskit/license.token`` file
        3. ``explicit_key`` parameter (if provided)

    Args:
        explicit_key: Optional explicit license key (lowest priority).

    Returns:
        Resolution result including source, token type, and validity.

    Raises:
        RuntimeError: If the C ABI call fails.
    """
    if explicit_key is not None:
        c_key = ffi.new("char[]", explicit_key.encode("utf-8"))
        ptr = lib.nxuskit_license_resolve(c_key)
    else:
        ptr = lib.nxuskit_license_resolve(ffi.NULL)

    if ptr == ffi.NULL:
        err = last_error() or "unknown error"
        raise RuntimeError(f"nxuskit_license_resolve failed: {err}")

    try:
        raw = json.loads(ffi.string(ptr).decode("utf-8"))
    finally:
        lib.nxuskit_free_string(ptr)

    return LicenseResolution(
        source=raw["source"],
        token_type=raw["token_type"],
        valid=raw["valid"],
        error=raw.get("error"),
    )


def license_validate(token: str) -> TokenInfo:
    """Validate a license token JWT string.

    Performs RS384 signature verification, claim parsing, and type-specific
    validation (expiry, machine binding, version ceiling).

    Args:
        token: The JWT token string to validate.

    Returns:
        Validation result including token type, edition, and expiry info.

    Raises:
        RuntimeError: If the C ABI call fails.
    """
    c_token = ffi.new("char[]", token.encode("utf-8"))
    ptr = lib.nxuskit_license_validate(c_token)

    if ptr == ffi.NULL:
        err = last_error() or "unknown error"
        raise RuntimeError(f"nxuskit_license_validate failed: {err}")

    try:
        raw = json.loads(ffi.string(ptr).decode("utf-8"))
    finally:
        lib.nxuskit_free_string(ptr)

    return TokenInfo(
        valid=raw["valid"],
        token_type=raw["token_type"],
        edition=raw.get("edition"),
        days_remaining=raw.get("days_remaining"),
        error=raw.get("error"),
        result=raw["result"],
    )


def license_machine_id() -> str:
    """Get the machine fingerprint for this device.

    Returns a ``sha256:<64-hex-chars>`` string derived from the OS machine ID.

    Returns:
        The machine fingerprint string.

    Raises:
        RuntimeError: If the machine ID cannot be determined
            (e.g., in Docker containers or minimal environments).
    """
    ptr = lib.nxuskit_license_machine_id()
    if ptr == ffi.NULL:
        err = last_error() or "machine ID unavailable"
        raise RuntimeError(f"license_machine_id failed: {err}")

    try:
        return ffi.string(ptr).decode("utf-8")
    finally:
        lib.nxuskit_free_string(ptr)


@dataclass
class ActivationResult:
    """Result of activating or deactivating a Pro license."""

    success: bool
    """Whether the operation succeeded"""
    seats_used: int
    """Number of machines currently using this license"""
    seats_total: int
    """Maximum number of machines allowed"""
    message: str
    """Human-readable status message"""
    error: Optional[str]
    """Error message if the operation failed"""


@dataclass
class TrialResult:
    """Result of a trial issuance or activation."""

    success: bool
    """Whether the operation succeeded"""
    days_remaining: int
    """Days until trial expiry"""
    message: str
    """Human-readable status message"""
    error: Optional[str]
    """Error message if the operation failed"""


def license_activate(purchase_id: str) -> ActivationResult:
    """Activate a Pro license on this machine.

    Calls the licensing microservice to validate the purchase ID, generate
    a machine-bound developer token, and store it locally.

    Args:
        purchase_id: The purchase ID received after buying Pro.

    Returns:
        Activation result including seat count and status.

    Raises:
        RuntimeError: If the C ABI call fails.
    """
    c_pid = ffi.new("char[]", purchase_id.encode("utf-8"))
    ptr = lib.nxuskit_license_activate(c_pid)

    if ptr == ffi.NULL:
        err = last_error() or "unknown error"
        raise RuntimeError(f"nxuskit_license_activate failed: {err}")

    try:
        raw = json.loads(ffi.string(ptr).decode("utf-8"))
    finally:
        lib.nxuskit_free_string(ptr)

    return ActivationResult(
        success=raw.get("success", True),
        seats_used=raw["seats_used"],
        seats_total=raw["seats_total"],
        message=raw["message"],
        error=raw.get("error"),
    )


def license_deactivate() -> ActivationResult:
    """Deactivate the Pro license on this machine.

    Releases this machine's seat and removes the stored token.

    Returns:
        Deactivation result including updated seat count.

    Raises:
        RuntimeError: If the C ABI call fails.
    """
    ptr = lib.nxuskit_license_deactivate()

    if ptr == ffi.NULL:
        err = last_error() or "unknown error"
        raise RuntimeError(f"nxuskit_license_deactivate failed: {err}")

    try:
        raw = json.loads(ffi.string(ptr).decode("utf-8"))
    finally:
        lib.nxuskit_free_string(ptr)

    return ActivationResult(
        success=raw.get("success", True),
        seats_used=raw["seats_used"],
        seats_total=raw["seats_total"],
        message=raw["message"],
        error=raw.get("error"),
    )


def license_trial_issue() -> TrialResult:
    """Issue a 30-day trial token for this machine.

    Returns:
        Trial issuance result including days remaining.

    Raises:
        RuntimeError: If the C ABI call fails.
    """
    ptr = lib.nxuskit_license_trial_issue()

    if ptr == ffi.NULL:
        err = last_error() or "unknown error"
        raise RuntimeError(f"nxuskit_license_trial_issue failed: {err}")

    try:
        raw = json.loads(ffi.string(ptr).decode("utf-8"))
    finally:
        lib.nxuskit_free_string(ptr)

    return TrialResult(
        success=raw.get("success", True),
        days_remaining=raw["days_remaining"],
        message=raw["message"],
        error=raw.get("error"),
    )


def license_trial_activate(code: str) -> TrialResult:
    """Activate a trial token (complete email verification).

    Args:
        code: The activation code received via email.

    Returns:
        Trial activation result.

    Raises:
        RuntimeError: If the C ABI call fails.
    """
    c_code = ffi.new("char[]", code.encode("utf-8"))
    ptr = lib.nxuskit_license_trial_activate(c_code)

    if ptr == ffi.NULL:
        err = last_error() or "unknown error"
        raise RuntimeError(f"nxuskit_license_trial_activate failed: {err}")

    try:
        raw = json.loads(ffi.string(ptr).decode("utf-8"))
    finally:
        lib.nxuskit_free_string(ptr)

    return TrialResult(
        success=raw.get("success", True),
        days_remaining=raw["days_remaining"],
        message=raw["message"],
        error=raw.get("error"),
    )

File: packages/nxuskit/src/license.rs
//! License management wrappers for the nxusKit SDK.
//!
//! Safe Rust API over the C ABI license functions: token resolution,
//! validation, and machine fingerprinting.

use crate::error::NxuskitError;
use crate::ffi;

/// Token resolution result from the license precedence chain.
#[derive(Debug, Clone, serde::Deserialize)]
pub struct LicenseResolution {
    /// Where the token was found: "env_var", "file", "api_param", or "none"
    pub source: String,
    /// Token type: "trial", "developer", "deployment", or "none"
    pub token_type: String,
    /// Whether the token passed validation
    pub valid: bool,
    /// Error message if validation failed
    pub error: Option<String>,
}

/// Token validation result.
#[derive(Debug, Clone, serde::Deserialize)]
pub struct TokenInfo {
    /// Whether the token is valid
    pub valid: bool,
    /// Token type: "trial", "developer", "deployment"
    pub token_type: String,
    /// Edition granted by the token
    pub edition: Option<String>,
    /// Days until token expiry (None for deployment tokens)
    pub days_remaining: Option<i64>,
    /// Error message if validation failed
    pub error: Option<String>,
    /// Entitlement result code
    pub result: String,
}

/// Resolve the active license token from all available sources.
///
/// Resolution order:
/// 1. `NXUSKIT_LICENSE_TOKEN` environment variable
/// 2. `~/.nxuskit/license.token` file
/// 3. `explicit_key` parameter (if provided)
///
/// Returns the resolution result including source, token type, and validity.
pub fn license_resolve(explicit_key: Option<&str>) -> Result<LicenseResolution, NxuskitError> {
    use std::ffi::{CStr, CString};

    let ek_cstr = explicit_key.map(|k| CString::new(k).unwrap());
    let ek_ptr = ek_cstr
        .as_ref()
        .map(|c| c.as_ptr())
        .unwrap_or(std::ptr::null());

    let ptr = ffi::ffi_call!(nxuskit_license_resolve, ek_ptr);

    if ptr.is_null() {
        return Err(NxuskitError::Internal {
            message: "nxuskit_license_resolve returned NULL".to_string(),
        });
    }

    let json_str = unsafe { CStr::from_ptr(ptr) }
        .to_str()
        .map_err(|e| NxuskitError::Internal {
            message: format!("license resolve not valid UTF-8: {e}"),
        })?
        .to_string();

    // Free the C string
    ffi::ffi_call!(nxuskit_free_string, ptr);

    serde_json::from_str(&json_str).map_err(|e| NxuskitError::Internal {
        message: format!("Failed to parse license resolve JSON: {e}"),
    })
}

/// Validate a license token JWT string.
///
/// Performs RS384 signature verification, claim parsing, and type-specific
/// validation (expiry, machine binding, version ceiling).
pub fn license_validate(token: &str) -> Result<TokenInfo, NxuskitError> {
    use std::ffi::{CStr, CString};

    let token_cstr = CString::new(token).map_err(|_| NxuskitError::Internal {
        message: "Token contains interior NUL byte".to_string(),
    })?;

    let ptr = ffi::ffi_call!(nxuskit_license_validate, token_cstr.as_ptr());

    if ptr.is_null() {
        return Err(NxuskitError::Internal {
            message: "nxuskit_license_validate returned NULL".to_string(),
        });
    }

    let json_str = unsafe { CStr::from_ptr(ptr) }
        .to_str()
        .map_err(|e| NxuskitError::Internal {
            message: format!("license validate not valid UTF-8: {e}"),
        })?
        .to_string();

    ffi::ffi_call!(nxuskit_free_string, ptr);

    serde_json::from_str(&json_str).map_err(|e| NxuskitError::Internal {
        message: format!("Failed to parse license validate JSON: {e}"),
    })
}

/// Get the machine fingerprint for this device.
///
/// Returns a `sha256:<64-hex-chars>` string derived from the OS machine ID.
///
/// Returns an error if the machine ID cannot be determined (e.g., in Docker
/// containers or minimal environments).
pub fn license_machine_id() -> Result<String, NxuskitError> {
    use std::ffi::CStr;

    let ptr = ffi::ffi_call!(nxuskit_license_machine_id);

    if ptr.is_null() {
        return Err(NxuskitError::Internal {
            message: "Machine ID unavailable (Docker container or minimal environment?)"
                .to_string(),
        });
    }

    let fingerprint = unsafe { CStr::from_ptr(ptr) }
        .to_str()
        .map_err(|e| NxuskitError::Internal {
            message: format!("machine ID not valid UTF-8: {e}"),
        })?
        .to_string();

    ffi::ffi_call!(nxuskit_free_string, ptr);

    Ok(fingerprint)
}

/// Activation result from the licensing microservice.
#[derive(Debug, Clone, serde::Deserialize)]
pub struct ActivationResult {
    pub success: bool,
    pub seats_used: u32,
    pub seats_total: u32,
    pub message: String,
    pub error: Option<String>,
}

/// Activate a Pro license on this machine.
///
/// Calls the licensing microservice to validate the purchase ID, generate
/// a machine-bound developer token, and store it locally.
pub fn license_activate(purchase_id: &str) -> Result<ActivationResult, NxuskitError> {
    use std::ffi::{CStr, CString};

    let pid_cstr = CString::new(purchase_id).map_err(|_| NxuskitError::Internal {
        message: "Purchase ID contains interior NUL byte".to_string(),
    })?;

    let ptr = ffi::ffi_call!(nxuskit_license_activate, pid_cstr.as_ptr());

    if ptr.is_null() {
        return Err(NxuskitError::Internal {
            message: "nxuskit_license_activate returned NULL".to_string(),
        });
    }

    let json_str = unsafe { CStr::from_ptr(ptr) }
        .to_str()
        .map_err(|e| NxuskitError::Internal {
            message: format!("activate result not valid UTF-8: {e}"),
        })?
        .to_string();

    ffi::ffi_call!(nxuskit_free_string, ptr);

    serde_json::from_str(&json_str).map_err(|e| NxuskitError::Internal {
        message: format!("Failed to parse activation JSON: {e}"),
    })
}

/// Deactivate the Pro license on this machine.
///
/// Releases this machine's seat and removes the stored token.
pub fn license_deactivate() -> Result<ActivationResult, NxuskitError> {
    use std::ffi::CStr;

    let ptr = ffi::ffi_call!(nxuskit_license_deactivate);

    if ptr.is_null() {
        return Err(NxuskitError::Internal {
            message: "nxuskit_license_deactivate returned NULL".to_string(),
        });
    }

    let json_str = unsafe { CStr::from_ptr(ptr) }
        .to_str()
        .map_err(|e| NxuskitError::Internal {
            message: format!("deactivate result not valid UTF-8: {e}"),
        })?
        .to_string();

    ffi::ffi_call!(nxuskit_free_string, ptr);

    serde_json::from_str(&json_str).map_err(|e| NxuskitError::Internal {
        message: format!("Failed to parse deactivation JSON: {e}"),
    })
}

/// Trial issuance result.
#[derive(Debug, Clone, serde::Deserialize)]
pub struct TrialResult {
    pub success: bool,
    pub days_remaining: u32,
    pub message: String,
    pub error: Option<String>,
}

/// Issue a 30-day trial token for this machine.
pub fn license_trial_issue() -> Result<TrialResult, NxuskitError> {
    use std::ffi::CStr;

    let ptr = ffi::ffi_call!(nxuskit_license_trial_issue);

    if ptr.is_null() {
        return Err(NxuskitError::Internal {
            message: "nxuskit_license_trial_issue returned NULL".to_string(),
        });
    }

    let json_str = unsafe { CStr::from_ptr(ptr) }
        .to_str()
        .map_err(|e| NxuskitError::Internal {
            message: format!("trial issue result not valid UTF-8: {e}"),
        })?
        .to_string();

    ffi::ffi_call!(nxuskit_free_string, ptr);

    serde_json::from_str(&json_str).map_err(|e| NxuskitError::Internal {
        message: format!("Failed to parse trial issue JSON: {e}"),
    })
}

/// Activate a trial token (complete email verification).
pub fn license_trial_activate(code: &str) -> Result<TrialResult, NxuskitError> {
    use std::ffi::{CStr, CString};

    let code_cstr = CString::new(code).map_err(|_| NxuskitError::Internal {
        message: "Activation code contains interior NUL byte".to_string(),
    })?;

    let ptr = ffi::ffi_call!(nxuskit_license_trial_activate, code_cstr.as_ptr());

    if ptr.is_null() {
        return Err(NxuskitError::Internal {
            message: "nxuskit_license_trial_activate returned NULL".to_string(),
        });
    }

    let json_str = unsafe { CStr::from_ptr(ptr) }
        .to_str()
        .map_err(|e| NxuskitError::Internal {
            message: format!("trial activate result not valid UTF-8: {e}"),
        })?
        .to_string();

    ffi::ffi_call!(nxuskit_free_string, ptr);

    serde_json::from_str(&json_str).map_err(|e| NxuskitError::Internal {
        message: format!("Failed to parse trial activate JSON: {e}"),
    })
}
